### PR TITLE
notifier(engine): fix Flush deadloop in closed notifier

### LINKED ISSUE #7506: server master leader gets stuck
### What did you do?

start two tiflow master servers

### What did you expect to see?

_No response_

### What did you see instead?

- executor register API call blocks (because of the RegisterExecutor API of server master doesn't return)
- server master hangs in executor manager

logs: 
[tiflow-cluster-logs.zip](https://github.com/pingcap/tiflow/files/9906497/tiflow-cluster-logs.zip)

goroutines: 
[goroutines.log](https://github.com/pingcap/tiflow/files/9906505/goroutines.log)

Block executor register
```go
goroutine 111653 [semacquire, 55 minutes]:
sync.runtime_SemacquireMutex(0xc0039d6bc0?, 0x1?, 0x1?)
    runtime/sema.go:77 +0x25
sync.(*Mutex).lockSlow(0xc0006788c0)
    sync/mutex.go:171 +0x165
sync.(*Mutex).Lock(...)
    sync/mutex.go:90
github.com/pingcap/tiflow/engine/servermaster.(*ExecutorManagerImpl).AllocateNewExec(0xc0006788a0, {0x4226a90, 0xc000f3c5a0}, 0xc000f3c960)
    github.com/pingcap/tiflow/engine/servermaster/executor_manager.go:186 +0x165
github.com/pingcap/tiflow/engine/servermaster.(*Server).RegisterExecutor(0xc000f6e240, {0x4226a90, 0xc000f3c5a0}, 0xc0032e36c8?)
    github.com/pingcap/tiflow/engine/servermaster/server.go:266 +0xf6
github.com/pingcap/tiflow/engine/enginepb._Discovery_RegisterExecutor_Handler.func1({0x4226a90, 0xc000f3c5a0}, {0x3522380?, 0xc000f3c960})
    github.com/pingcap/tiflow/engine/enginepb/master_grpc.pb.go:211 +0x78
github.com/pingcap/tiflow/engine/pkg/rpcerror.UnaryServerInterceptor({0x4226a90?, 0xc000f3c5a0?}, {0x3522380, 0xc000f3c960}, 0xc00399aec0, 0xc0039e450
    github.com/pingcap/tiflow/engine/pkg/rpcerror/grpc.go:140 +0x5a
google.golang.org/grpc.chainUnaryInterceptors.func1.1({0x4226a90?, 0xc000f3c5a0?}, {0x3522380?, 0xc000f3c960?})
    google.golang.org/grpc@v1.48.0/server.go:1129 +0x5b
github.com/grpc-ecosystem/go-grpc-prometheus.(*ServerMetrics).UnaryServerInterceptor.func1({0x4226a90, 0xc000f3c5a0}, {0x3522380, 0xc000f3c960}, 0x361
    github.com/grpc-ecosystem/go-grpc-prometheus@v1.2.0/server_metrics.go:107 +0x87
google.golang.org/grpc.chainUnaryInterceptors.func1.1({0x4226a90?, 0xc000f3c5a0?}, {0x3522380?, 0xc000f3c960?})
    google.golang.org/grpc@v1.48.0/server.go:1132 +0x83
google.golang.org/grpc.chainUnaryInterceptors.func1({0x4226a90, 0xc000f3c5a0}, {0x3522380, 0xc000f3c960}, 0xc00399aec0, 0xc001367710)
    google.golang.org/grpc@v1.48.0/server.go:1134 +0x12b
github.com/pingcap/tiflow/engine/enginepb._Discovery_RegisterExecutor_Handler({0x37fac20?, 0xc000f6e240}, {0x4226a90, 0xc000f3c5a0}, 0xc000428070, 0xc
    github.com/pingcap/tiflow/engine/enginepb/master_grpc.pb.go:213 +0x138
google.golang.org/grpc.(*Server).processUnaryRPC(0xc000eceb40, {0x424af40, 0xc0032ca000}, 0xc001ba6b40, 0xc0013f8450, 0x67ef700, 0x0)
    google.golang.org/grpc@v1.48.0/server.go:1295 +0xb2b
google.golang.org/grpc.(*Server).handleStream(0xc000eceb40, {0x424af40, 0xc0032ca000}, 0xc001ba6b40, 0x0)
    google.golang.org/grpc@v1.48.0/server.go:1636 +0xa2f
google.golang.org/grpc.(*Server).serveStreams.func1.2()
    google.golang.org/grpc@v1.48.0/server.go:932 +0x98
created by google.golang.org/grpc.(*Server).serveStreams.func1
    google.golang.org/grpc@v1.48.0/server.go:930 +0x28a
```

Root cause 

notifier.Flush enters a deadloop.

```go
goroutine 19833 [runnable]:
github.com/pingcap/tiflow/engine/pkg/notifier.(*Notifier[...]).Flush(0xc000678840, {0x42269e8, 0xc001972ac0})
    github.com/pingcap/tiflow/engine/pkg/notifier/notifier.go:137 +0x85
github.com/pingcap/tiflow/engine/servermaster.(*ExecutorManagerImpl).WatchExecutors(0xc0006788a0, {0x42269e8, 0xc001972ac0})
    github.com/pingcap/tiflow/engine/servermaster/executor_manager.go:401 +0x1e5
github.com/pingcap/tiflow/engine/pkg/externalresource/manager.(*DefaultGCCoordinator).initializeGC(0xc001972a80, {0x42269e8, 0xc001972ac0})
    github.com/pingcap/tiflow/engine/pkg/externalresource/manager/gc_coordinator.go:147 +0xba
github.com/pingcap/tiflow/engine/pkg/externalresource/manager.(*DefaultGCCoordinator).Run(0x0?, {0x42269e8, 0xc001972ac0})
    github.com/pingcap/tiflow/engine/pkg/externalresource/manager/gc_coordinator.go:74 +0xc7
github.com/pingcap/tiflow/engine/servermaster.(*Server).runLeaderService.func10()
    github.com/pingcap/tiflow/engine/servermaster/server.go:882 +0x34
golang.org/x/sync/errgroup.(*Group).Go.func1()
    golang.org/x/sync@v0.0.0-20220722155255-886fb9371eb4/errgroup/errgroup.go:75 +0x64
created by golang.org/x/sync/errgroup.(*Group).Go
    golang.org/x/sync@v0.0.0-20220722155255-886fb9371eb4/errgroup/errgroup.go:72 +0xa5
```

```go
goroutine 19800 [select, 57 minutes]:
github.com/pingcap/tiflow/engine/pkg/notifier.(*Notifier[...]).run(0xc0018c4d20)
    github.com/pingcap/tiflow/engine/pkg/notifier/notifier.go:160 +0x12c
github.com/pingcap/tiflow/engine/pkg/notifier.NewNotifier[...].func1()
    github.com/pingcap/tiflow/engine/pkg/notifier/notifier.go:89 +0x66
created by github.com/pingcap/tiflow/engine/pkg/notifier.NewNotifier[...]
    github.com/pingcap/tiflow/engine/pkg/notifier/notifier.go:87 +0x185
```

### Versions of the cluster

Dataflow Engine version (run `tiflow version`):

[INFO] [version.go:47] ["Welcome to TiFlow Master"] [release-version=v6.3.0-master] [git-hash=b5b4d3d22d36ec40a0de086db23eb49db42c256a] [git-branch=detached] [utc-build-time="2022-10-27 06:07:08"] [go-version="go version go1.19.2 linux/amd64"] [failpoint-build=false]

### PR DESCRIPTION
**Pull request by amyangfei**
### What problem does this PR solve?

Issue Number: close #7506

### What is changed and how it works?

- Flush should return immediately if notifier has been closed.

#### Tests 

 - Unit test
 - Integration test

### Release note 

```release-note
None
```

### CODE REVIEW

File: engine/servermaster/executor_manager.go [hunk 1]
```diff
@@ -282,40 +282,33 @@ func (e *Executor) statusEqual(status model.ExecutorStatus) bool {
...
 			}
 		}
-	}()
-
-	return nil
+	}
 }
 
 // Stop implements ExecutorManager.Stop
 func (e *ExecutorManagerImpl) Stop() {
```
Comment (gozssky) [inline engine/servermaster/executor_manager.go:311]: We can remove this method and inline `e.notifier.Close()` to `Run` method.

**Comment by amyangfei** [comment]
run-engine-integration-test

**Review by gozssky — APPROVED** [review]

**Review by CharlesCheung96 — APPROVED** [review]

**Pull request merged**